lib/puppet/pops/evaluator/runtime3_support.rb

```diff
@@ -33,14 +33,6 @@ def fail(issue, semantic, options={}, except=nil)
   # @raise [Puppet::ParseError] an evaluation error initialized from the arguments (TODO: Change to EvaluationError?)
   #
   def optionally_fail(issue, semantic, options={}, except=nil)
-    if except.nil?
-      # Want a stacktrace, and it must be passed as an exception
-      begin
-       raise EvaluationError.new()
-      rescue EvaluationError => e
-        except = e
-      end
-    end
     diagnostic_producer.accept(issue, semantic, options, except)
```

Comment:
It looks like `optionally_fail` is also called by `fail` and the latter defaults the `except` to nil as well. It makes me a little nervous that previously we never passed a `nil` exception, but now we are. Would it make sense to continue to create an exception and set the backtrace, but not raise, something like the following?

```ruby
if except.nil?
  except = EvaluationError.new
  except.set_backtrace(caller)
end
diagnostic_producer.accept(issue, semantic, options, except)
```

Reply:
I can try it out and see, though I think throwing exceptions and calling `caller()` are slow for similar reasons.

Reply:
fwiw, I started to go through what `optionally_fail` does and the exception is unused until it's passed into [Diagnostic.new](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/validation.rb#L216), whose initialize method has [this gem](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/validation.rb#L237-L239). I tried removing the `exception` reader from `Diagnostic` but it caused hundreds of test failures - so that comment is probably wrong...

Reply:
Okay, double checking that I'm understanding correctly:
After accepting these diagnostics the IssueReporter will either [create a new exception with this backtrace](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/issue_reporter.rb#L55-L58), or [log each diagnostic at error level](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/issue_reporter.rb#L67), depending on the number of errors and the max_errors value.

The above code will use the diagnostic's `exception` reader. And we want the exception's backtrace to point to `optionally_fail` rather than IssueReporter because optionally_fail will be closer to the point actually being parsed (Runtime3Support is `include`d in the Compiler, Evaluator, data converter, and some operators.

However, `optionally_fail` will only actually fail if we have the `strict` setting == `:error` (default `:warning`) & the setting `trace` == true (default true in server).

Reply:
I'd love to just have `fail` create its exception/callstack prior to calling `optionally_fail` and pass it through. However, for optional failures, it would seem ideal to ask the [`SeverityProducer`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/evaluator/runtime3_support.rb#L481) if it was a failure before creating the callstack.

Reply:
I think something like:
```rb
if except.nil? && diagnostic_producer.severity_producer[issue] == :error
  except = EvaluationError.new
  except.set_backtrace(caller)
end
diagnostic_producer.accept(issue, semantic, options, except)
```

Would give us what we want??